src/loader.js

```diff
@@ -0,0 +1,159 @@
+let util = require('util');
+let assume = require('assume');
+let debug = require('debug')('taskcluster:loader');
+let Promise = require('promise');
+let _ = require('lodash');
+let TopoSort = require('topo-sort');
+
+// We use this closure to create a .create() method which
+// returns a graphViz file for a given component directory
+function generateGraph(dir, vComp) {
+  let tsort = new TopoSort();
+
+  for (let x of _.keys(dir)) {
+    tsort.add(x, dir[x].requires || []);
+  }
+
+  for (let x of _.keys(vComp)) {
+    tsort.add(x, vComp[x].requires || []);
+  }
+
+  let components = tsort.sort();
+
+  let dot = [
+    '// This graph shows all dependencies for this loader',
+    '// including virtual dependencies.',
+    '// You might find http://www.webgraphviz.com/ useful!',
+    '',
+    'digraph G {',
+  ];
+
+  for (let component of components) {
+    dot.push(util.format('  "%s"', component));
+    let node = dir[component] || vComp[component];
+    for (let dependency of node.requires || []) {
+      dot.push(util.format('  "%s" -> "%s" [dir=back]', component, dependency));
+    }
+  }
+  dot.push('}');
+
+  dot = dot.join('\n');
+
+  return dot;
+}
+
+function loader(componentDirectory, requiredVirtualComponents = []) {
+  assume(componentDirectory).is.an('object');
+  assume(requiredVirtualComponents).is.an('array');
+  assume(_.intersection(_.keys(componentDirectory), requiredVirtualComponents)).has.length(0);
+
+  let knownComponents = _.keys(componentDirectory).concat(requiredVirtualComponents);
+
+  return function(virtualComponents = {}) {
+    assume(virtualComponents).is.an('object');
+
+    let graphViz;
+    try {
+      graphViz = generateGraph(componentDirectory, virtualComponents);
+    } catch (err) {
+      if (err.message.match(/^At least \d+ circular dependency in nodes/)) {
+        let errStr = util.format('Cyclical dependency: %s', err.message);
+        debug(errStr);
+        throw new Error(errStr);
+      }
+      throw err;
```

Comment:
So, the ideal case is base components are complex and virtual ones are flat.  That's great, but what happens when in testing we need, e.g.  the table name to be generated based on an async call to some web service which depends on an api we only configure for testing.

Whether we have a rule to enforce the ideal case or not, I'd like to restrict the code internally to that.